src/test/client/nonblocking.cc

```diff
@@ -611,3 +612,61 @@ TEST_F(TestClient, LlreadvLlwritevInvalidFileHandle) {
   ASSERT_EQ(bytes_read, -CEPHFS_EBADF);
   ASSERT_EQ(bl.length(), 0);
 }
+
+TEST_F(TestClient, LlreadvLlwritevLargeBuffers) {
+  /* Test that async I/O code paths handle large buffers (total len >= 4GiB)*/
+  int mypid = getpid();
+  char filename[256];
+
+  client->unmount();
+  TearDown();
+  SetUp();
+
+  sprintf(filename, "test_llreadvllwritevlargebuffers%u", mypid);
+
+  Inode *root, *file;
+  root = client->get_root();
+  ASSERT_NE(root, (Inode *)NULL);
+
+  Fh *fh;
+  struct ceph_statx stx;
+
+  ASSERT_EQ(0, client->ll_createx(root, filename, 0666,
+                                  O_RDWR | O_CREAT | O_TRUNC,
+                                          &file, &fh, &stx, 0, 0, myperm));
+
+  struct statvfs stbuf;
+  int64_t rc;
+  rc = client->ll_statfs(root, &stbuf, myperm);
+  ASSERT_EQ(rc, 0);
+  int64_t fs_available_space = stbuf.f_bfree * stbuf.f_bsize;
+  ASSERT_GT(fs_available_space, 0);
+
+  const int64_t BUFSIZE = 1024 * 1024 * 1024;
+  int64_t bytes_written = 0;
+
+  std::unique_ptr<C_SaferCond> writefinish = nullptr;
+  writefinish.reset(new C_SaferCond("test-nonblocking-writefinish-large-buffers"));
+
+  auto out_buf_0 = std::make_unique<char[]>(BUFSIZE * 2);
+  memset(out_buf_0.get(), 0xDD, BUFSIZE * 2);
+  auto out_buf_1 = std::make_unique<char[]>(BUFSIZE * 2);
+  memset(out_buf_1.get(), 0xFF, BUFSIZE * 2);
+
+  struct iovec iov_out[2] = {
+    {out_buf_0.get(), BUFSIZE * 2},
+    {out_buf_1.get(), BUFSIZE * 2}
+  };
+
+  bufferlist bl;
+
+  rc = client->ll_preadv_pwritev(fh, iov_out, 2, 0, true, writefinish.get(),
+                                 nullptr);
```

Comment:
Is this just contained to async writes? I tried using `ll_writev` and am running into same symptom. 

This is what I changed. Can you chase this down?
```suggestion
 rc = client->ll_writev(fh, iov_out, 2, 0);
```

Reply:
(Yikes) Let me take a look

Reply:
something's not right with `CFG_block_writes_upfront`

Reply:
I'll debug more on this tomorrow

Reply:
@chrisphoffman the issue is because the sync read/write paths (ll_{read|write}v) are not clamping the i/o size.